flang/lib/Evaluate/check-expression.cpp

```diff
@@ -132,7 +132,7 @@ bool IsConstantExprHelper<INVARIANT>::operator()(
     } else if (intrinsic->name == "shape" || intrinsic->name == "size") {
       auto shape{GetShape(call.arguments()[0]->UnwrapExpr())};
       return shape && IsConstantExprShape(*shape);
-    } else if (proc.IsPure()) {
+    } else if (proc.IsPure() || proc.IsSimple()) {
```

Comment:
All simple procedures are also pure, so you don't have to check for `simple` when you know it's not `pure`.